File: backend/app/routers/trade.py
# app/routers/trade.py
from typing import Annotated, List, Optional
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.dependencies import get_current_approved_user
from app.models.user import User
from app.schemas.trade import (
    TradeRunCreate,
    TradeRunUpdate,
    TradeRunResponse,
    PriceReportCreate,
    PriceReportResponse,
    CargoContractCreate,
    CargoContractUpdate,
    CargoContractResponse,
)
from app.services.trade import TradeService

router = APIRouter(prefix="/api/trade", tags=["trade"])


# === Trade Run Endpoints ===

@router.post("/runs", response_model=TradeRunResponse, status_code=status.HTTP_201_CREATED)
async def create_trade_run(
    data: TradeRunCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Log a new trade run."""
    service = TradeService(db)
    trade_run = await service.create_trade_run(current_user.id, data)
    return trade_run


@router.get("/runs", response_model=List[TradeRunResponse])
async def get_my_trade_runs(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """Get the current user's trade runs."""
    service = TradeService(db)
    return await service.get_user_trade_runs(current_user.id, skip=skip, limit=limit)


@router.get("/runs/all", response_model=List[TradeRunResponse])
async def get_all_trade_runs(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """Get all trade runs (for leaderboard context)."""
    service = TradeService(db)
    return await service.get_all_trade_runs(skip=skip, limit=limit)


@router.get("/runs/{run_id}", response_model=TradeRunResponse)
async def get_trade_run(
    run_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get a specific trade run."""
    service = TradeService(db)
    trade_run = await service.get_trade_run_by_id(run_id)
    if not trade_run:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Trade run not found"
        )
    return trade_run


@router.put("/runs/{run_id}", response_model=TradeRunResponse)
async def update_trade_run(
    run_id: int,
    data: TradeRunUpdate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Update a trade run."""
    service = TradeService(db)
    trade_run = await service.get_trade_run_by_id(run_id)
    
    if not trade_run:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Trade run not found"
        )
    
    if trade_run.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Can only edit your own trade runs"
        )
    
    updated = await service.update_trade_run(trade_run, data)
    return updated


@router.delete("/runs/{run_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_trade_run(
    run_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Delete a trade run."""
    service = TradeService(db)
    trade_run = await service.get_trade_run_by_id(run_id)
    
    if not trade_run:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Trade run not found"
        )
    
    if trade_run.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Can only delete your own trade runs"
        )
    
    await service.delete_trade_run(trade_run)


@router.get("/stats/my", response_model=dict)
async def get_my_trade_stats(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get the current user's trading statistics."""
    service = TradeService(db)
    return await service.get_user_trade_stats(current_user.id)


@router.get("/leaderboard", response_model=List[dict])
async def get_leaderboard(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    limit: int = Query(10, ge=1, le=50),
):
    """Get the trading leaderboard."""
    service = TradeService(db)
    return await service.get_leaderboard(limit=limit)


# === Price Report Endpoints ===

@router.post("/prices", response_model=PriceReportResponse, status_code=status.HTTP_201_CREATED)
async def create_price_report(
    data: PriceReportCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Submit a new price report."""
    service = TradeService(db)
    price_report = await service.create_price_report(current_user.id, data)
    return price_report


@router.get("/prices", response_model=List[PriceReportResponse])
async def get_price_reports(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    location: Optional[str] = Query(None),
    commodity: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=200),
):
    """Get price reports with optional filtering."""
    service = TradeService(db)
    return await service.get_price_reports(
        location=location,
        commodity=commodity,
        skip=skip,
        limit=limit
    )


@router.get("/prices/location/{location}", response_model=List[PriceReportResponse])
async def get_prices_by_location(
    location: str,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get latest prices for all commodities at a location."""
    service = TradeService(db)
    return await service.get_latest_prices_by_location(location)


@router.get("/prices/history", response_model=List[PriceReportResponse])
async def get_price_history(
    location: str = Query(..., description="Location name"),
    commodity: str = Query(..., description="Commodity name"),
    days: int = Query(7, ge=1, le=30),
    db: Annotated[AsyncSession, Depends(get_db)] = None,
    current_user: Annotated[User, Depends(get_current_approved_user)] = None,
):
    """Get price history for a commodity at a location."""
    service = TradeService(db)
    return await service.get_price_history(location, commodity, days)


@router.get("/routes/best", response_model=List[dict])
async def get_best_routes(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    limit: int = Query(10, ge=1, le=20),
):
    """Get the best trade routes based on current price data."""
    service = TradeService(db)
    return await service.get_best_routes(limit=limit)


# === Cargo Contract Endpoints ===

@router.post("/contracts", response_model=CargoContractResponse, status_code=status.HTTP_201_CREATED)
async def create_cargo_contract(
    data: CargoContractCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Post a new cargo contract."""
    service = TradeService(db)
    contract = await service.create_cargo_contract(current_user.id, data)
    return contract


@router.get("/contracts/open", response_model=List[CargoContractResponse])
async def get_open_contracts(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """Get all open cargo contracts."""
    service = TradeService(db)
    return await service.get_open_contracts(skip=skip, limit=limit)


@router.get("/contracts/my", response_model=List[CargoContractResponse])
async def get_my_contracts(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get contracts posted by the current user."""
    service = TradeService(db)
    return await service.get_posted_contracts(current_user.id)


@router.get("/contracts/hauling", response_model=List[CargoContractResponse])
async def get_my_hauling_contracts(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get contracts the current user has accepted to haul."""
    service = TradeService(db)
    return await service.get_accepted_contracts(current_user.id)


@router.get("/contracts/{contract_id}", response_model=CargoContractResponse)
async def get_contract(
    contract_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get a specific cargo contract."""
    service = TradeService(db)
    contract = await service.get_contract_by_id(contract_id)
    
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    return contract


@router.post("/contracts/{contract_id}/accept", response_model=CargoContractResponse)
async def accept_contract(
    contract_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Accept a cargo contract as a hauler."""
    service = TradeService(db)
    contract = await service.get_contract_by_id(contract_id)
    
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    try:
        updated = await service.accept_contract(contract, current_user.id)
        return updated
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/contracts/{contract_id}/start", response_model=CargoContractResponse)
async def start_contract(
    contract_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Mark a contract as in progress. Only the hauler can do this."""
    service = TradeService(db)
    contract = await service.get_contract_by_id(contract_id)
    
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    if contract.hauler_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only the hauler can start the contract"
        )
    
    try:
        updated = await service.start_contract(contract)
        return updated
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/contracts/{contract_id}/complete", response_model=CargoContractResponse)
async def complete_contract(
    contract_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Complete a cargo contract. Only the poster or hauler can do this."""
    service = TradeService(db)
    contract = await service.get_contract_by_id(contract_id)
    
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    if contract.poster_id != current_user.id and contract.hauler_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only poster or hauler can complete"
        )
    
    try:
        updated = await service.complete_contract(contract)
        return updated
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/contracts/{contract_id}/cancel", response_model=CargoContractResponse)
async def cancel_contract(
    contract_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Cancel a cargo contract. Only the poster or hauler can do this."""
    service = TradeService(db)
    contract = await service.get_contract_by_id(contract_id)
    
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    try:
        updated = await service.cancel_contract(contract, current_user.id)
        return updated
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/contracts/{contract_id}/dispute", response_model=CargoContractResponse)
async def dispute_contract(
    contract_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Mark a contract as disputed for admin review."""
    service = TradeService(db)
    contract = await service.get_contract_by_id(contract_id)
    
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    if contract.poster_id != current_user.id and contract.hauler_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only involved parties can dispute"
        )
    
    try:
        updated = await service.dispute_contract(contract)
        return updated
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/hauler/stats", response_model=dict)
async def get_hauler_stats(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_approved_user)],
):
    """Get hauling statistics for the current user."""
    service = TradeService(db)
    return await service.get_hauler_stats(current_user.id)

File: frontend/src/stores/trade.js
import { defineStore } from 'pinia';
import api from '../services/api';

export const useTradeStore = defineStore('trade', {
  state: () => ({
    tradeRuns: [],
    priceReports: [],
    contracts: [],
    myContracts: [],
    myHaulingContracts: [],
    tradeStats: null,
    haulerStats: null,
    leaderboard: [],
    bestRoutes: [],
    isLoading: false,
    error: null,
  }),
  actions: {
    // Trade Runs
    async fetchTradeRuns() {
      this.isLoading = true;
      try {
        const response = await api.get('/trade/runs');
        this.tradeRuns = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch trade runs';
      } finally {
        this.isLoading = false;
      }
    },

    async createTradeRun(runData) {
      try {
        const response = await api.post('/trade/runs', runData);
        this.tradeRuns.unshift(response.data);
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to create trade run';
        throw err;
      }
    },

    async deleteTradeRun(runId) {
      try {
        await api.delete(`/trade/runs/${runId}`);
        this.tradeRuns = this.tradeRuns.filter(r => r.id !== runId);
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to delete trade run';
        throw err;
      }
    },

    async fetchTradeStats() {
      try {
        const response = await api.get('/trade/stats/my');
        this.tradeStats = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch trade stats';
      }
    },

    async fetchLeaderboard() {
      try {
        const response = await api.get('/trade/leaderboard');
        this.leaderboard = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch leaderboard';
      }
    },

    // Price Reports
    async fetchPriceReports(filters = {}) {
      this.isLoading = true;
      try {
        const params = new URLSearchParams();
        if (filters.location) params.append('location', filters.location);
        if (filters.commodity) params.append('commodity', filters.commodity);
        
        const response = await api.get(`/trade/prices?${params}`);
        this.priceReports = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch price reports';
      } finally {
        this.isLoading = false;
      }
    },

    async createPriceReport(reportData) {
      try {
        const response = await api.post('/trade/prices', reportData);
        this.priceReports.unshift(response.data);
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to create price report';
        throw err;
      }
    },

    async fetchBestRoutes() {
      try {
        const response = await api.get('/trade/routes/best');
        this.bestRoutes = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch best routes';
      }
    },

    // Cargo Contracts
    async fetchOpenContracts() {
      this.isLoading = true;
      try {
        const response = await api.get('/trade/contracts/open');
        this.contracts = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch contracts';
      } finally {
        this.isLoading = false;
      }
    },

    async fetchMyContracts() {
      try {
        const response = await api.get('/trade/contracts/my');
        this.myContracts = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch my contracts';
      }
    },

    async fetchMyHaulingContracts() {
      try {
        const response = await api.get('/trade/contracts/hauling');
        this.myHaulingContracts = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch hauling contracts';
      }
    },

    async createContract(contractData) {
      try {
        const response = await api.post('/trade/contracts', contractData);
        this.myContracts.unshift(response.data);
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to create contract';
        throw err;
      }
    },

    async acceptContract(contractId) {
      try {
        const response = await api.post(`/trade/contracts/${contractId}/accept`);
        // Update contract in lists
        const idx = this.contracts.findIndex(c => c.id === contractId);
        if (idx !== -1) {
          this.contracts[idx] = response.data;
        }
        this.myHaulingContracts.unshift(response.data);
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to accept contract';
        throw err;
      }
    },

    async completeContract(contractId) {
      try {
        const response = await api.post(`/trade/contracts/${contractId}/complete`);
        // Update in all lists
        [this.contracts, this.myContracts, this.myHaulingContracts].forEach(list => {
          const idx = list.findIndex(c => c.id === contractId);
          if (idx !== -1) list[idx] = response.data;
        });
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to complete contract';
        throw err;
      }
    },

    // NEW: startContract action
    async startContract(contractId) {
      try {
        const response = await api.post(`/trade/contracts/${contractId}/start`);
        // Update in all lists
        [this.contracts, this.myContracts, this.myHaulingContracts].forEach(list => {
          const idx = list.findIndex(c => c.id === contractId);
          if (idx !== -1) list[idx] = response.data;
        });
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to start contract';
        throw err;
      }
    },

    async cancelContract(contractId) {
      try {
        const response = await api.post(`/trade/contracts/${contractId}/cancel`);
        // Update in all lists
        [this.contracts, this.myContracts, this.myHaulingContracts].forEach(list => {
          const idx = list.findIndex(c => c.id === contractId);
          if (idx !== -1) list[idx] = response.data;
        });
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to cancel contract';
        throw err;
      }
    },

    async fetchHaulerStats() {
      try {
        const response = await api.get('/trade/hauler/stats');
        this.haulerStats = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch hauler stats';
      }
    },
  }
});
